File: VTI/Services/TestGeneratorService.swift
import Foundation

struct TestGeneratorService: Sendable {
    let repositoryService: RepositoryService
    let latexService: LaTeXService
    let pdfCompilerService: PDFCompilerService

    func generate(
        title: String,
        date: Date,
        quesiti: [Quesito]
    ) throws -> GeneratedDocument {
        let destinationFolder = try repositoryService.testFolder()
        let timestamp = Self.timestampFormatter.string(from: Date())
        let safeTitle = sanitizedSegment(title.isEmpty ? "Test" : title)
        let dateString = Self.fileDateFormatter.string(from: date)
        let safeDate = sanitizedSegment(dateString)
        let fileName = "Test_\(safeTitle)_\(safeDate)_\(timestamp)"
        let texURL = destinationFolder.appendingPathComponent(fileName).appendingPathExtension("tex")

        let body = quesiti.map(\.latexBlock).joined(separator: "\n\n")
        _ = try latexService.writeRenderedTemplate(
            template: .test,
            placeholders: [
                "TITLE": title.isEmpty ? "Test personalizzato" : title,
                "DATE": Self.displayDateFormatter.string(from: date),
                "BODY": body
            ],
            destinationURL: texURL
        )

        do {
            let pdf = try pdfCompilerService.compile(texURL: texURL)
            return GeneratedDocument(texURL: texURL, pdfURL: pdf.pdfURL, log: pdf.log)
        } catch let error as AppError {
            return GeneratedDocument(texURL: texURL, pdfURL: nil, log: error.localizedDescription)
        }
    }

    private func sanitizedSegment(_ value: String) -> String {
        value
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .replacingOccurrences(of: " ", with: "_")
            .components(separatedBy: CharacterSet.alphanumerics.inverted)
            .filter { !$0.isEmpty }
            .joined(separator: "_")
    }

    private static let timestampFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyyMMdd_HHmmss"
        return formatter
    }()

    private static let fileDateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter
    }()

    private static let displayDateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateStyle = .medium
        formatter.timeStyle = .none
        formatter.locale = Locale(identifier: "it_IT")
        return formatter
    }()
}

File: VTI.Windows/src/VTI.Core/TestGeneratorService.cs
using System.Text.RegularExpressions;

namespace VTI.Core;

public sealed class TestGeneratorService
{
    private readonly RepositoryService _repository;
    private readonly LaTeXService _latex;
    private readonly PDFCompilerService _pdf;

    public TestGeneratorService(RepositoryService repository, LaTeXService latex, PDFCompilerService pdf)
    {
        _repository = repository;
        _latex = latex;
        _pdf = pdf;
    }

    public GeneratedDocument Generate(string title, DateTime date, IReadOnlyList<Quesito> quesiti)
    {
        var destinationFolder = _repository.TestFolder();
        Directory.CreateDirectory(destinationFolder);
        var timestamp = DateTime.Now.ToString("yyyyMMdd_HHmmss");
        var safeTitle = SanitizedSegment(string.IsNullOrWhiteSpace(title) ? "Test" : title);
        var dateString = date.ToString("yyyy-MM-dd");
        var safeDate = SanitizedSegment(dateString);
        var fileName = $"Test_{safeTitle}_{safeDate}_{timestamp}.tex";
        var texPath = Path.Combine(destinationFolder, fileName);
        var list = quesiti.ToList();
        var body = string.Join("\n\n", list.Select(q => q.LatexBlock));
        var solutions = SolutionsAppendixBuilder.BuildLongTable(list);
        var displayDate = date.ToString("d", new System.Globalization.CultureInfo("it-IT"));
        _latex.WriteRenderedTemplate(LaTeXTemplate.Test, new Dictionary<string, string>
        {
            ["TITLE"] = string.IsNullOrWhiteSpace(title) ? "Test personalizzato" : title,
            ["DATE"] = displayDate,
            ["BODY"] = body,
            ["SOLUTIONS"] = solutions
        }, texPath);

        try
        {
            var pdf = _pdf.Compile(texPath);
            return new GeneratedDocument { TexPath = texPath, PdfPath = pdf.PdfPath, Log = pdf.Log };
        }
        catch (AppException ex)
        {
            return new GeneratedDocument { TexPath = texPath, PdfPath = null, Log = ex.Message };
        }
    }

    private static string SanitizedSegment(string value)
    {
        var s = Regex.Replace(value.Trim().Replace(' ', '_'), @"[^a-zA-Z0-9]+", "_");
        s = s.Trim('_');
        return string.IsNullOrEmpty(s) ? "Test" : s;
    }
}
